### ISSUE TITLE
Allow setting the schema for an entire migration class

### ISSUE THREAD

**Issue by dennisdoomen** [fluentmigrator/fluentmigrator#290]
Currently, you have to repeatingly call InSchema() in every statement. It would be nice to be able to associate the migration class with a particular schema. E.g. by overriding the Schema property from the MigrationBase.

**Comment by monnster** [comment]
As I see there's still no solution, so for those who's looking for a workaround on this issue:
You will need to use some AOP library, I used LightInject.Interception.

First create a base class for your migrations and replace Create property so that it returns a proxy with our interceptor:
```csharp
public abstract class DatabaseMigrationBase: Migration
{
	public static readonly string DefaultSchema;
	private IServiceContainer _container;

	static DatabaseMigrationBase()
	{
		DefaultSchema = "my";
	}

	protected DatabaseMigrationBase()
	{
		_container = new ServiceContainer();

		_container.Register<ICreateExpressionRoot>(_ => BaseCreate(), new PerContainerLifetime());

		_container.Intercept(
			sr => sr.ServiceType == typeof(ICreateExpressionRoot), 
			sf => new SetDefaultSchemaInterceptor(DefaultSchema));
	}

	/// <summary>
	/// Replace base method so that proxy with our interceptor is returned.
	/// </summary>
	public new ICreateExpressionRoot Create
	{
		get { return _container.GetInstance<ICreateExpressionRoot>(); }
	}

	private ICreateExpressionRoot BaseCreate()
	{
		return base.Create;
	}
}
```

Now implement interceptor which checks every call of `ICreateCompositionExpressionRoot` members and searches for InSchema methods in its results:

```csharp
public class SetDefaultSchemaInterceptor: IInterceptor
{
	private string _defaultSchema;

	public SetDefaultSchemaInterceptor(string defaultSchema)
	{
		_defaultSchema = defaultSchema;
	}

	public object Invoke(IInvocationInfo invocationInfo)
	{
		// execute any method of ICreateExpressionRoot and get it's result
		var result = invocationInfo.Proceed();

		// skip methods with void return types
		if (null == result || string.IsNullOrEmpty(_defaultSchema))
			return result;

		// get all methods named InSchema
		var inSchemaMethods = result.GetType()
				.GetMethods()
				// add interface members implemented by type explicitly
				.Union(GetExplicitMethods(result.GetType()))
				.Where(m => m.Name == "InSchema" || m.Name.EndsWith(".InSchema"));

		foreach (var method in inSchemaMethods)
		{
			// check signature - one string parameter
			var parameters = method.GetParameters();
			if (parameters.Count() == 1 && parameters.First().ParameterType == typeof (string))
			{
				// if ok - set schema
				method.Invoke(result, new object[] { _defaultSchema });
			}
		}

		return result;
	}

	private MethodInfo[] GetExplicitMethods(Type type)
	{
		return type
			.GetInterfaces()
			.SelectMany(t => type.GetInterfaceMap(t).InterfaceMethods)
			.ToArray();
	}
}
```

**Comment by SlyNet** [comment]
Why does migrator is ignoring the `Search Path` option from the connection string? Is it an intended behavior?

**Comment by fubar-coder** [comment]
Fixed by #772
